src/source/source_state.js

```diff
@@ -16,50 +16,137 @@ export type LayerFeatureStates = {[layer: string]: FeatureStates};
 class SourceFeatureState {
     state: LayerFeatureStates;
     stateChanges: LayerFeatureStates;
+    deletedStates: {};
 
     constructor() {
         this.state = {};
         this.stateChanges = {};
+        this.deletedStates = {};
     }
 
-    updateState(sourceLayer: string, featureId: number, state: Object) {
+    updateState(sourceLayer: string, featureId: number, newState: Object) {
         const feature = String(featureId);
         this.stateChanges[sourceLayer] = this.stateChanges[sourceLayer] || {};
         this.stateChanges[sourceLayer][feature] = this.stateChanges[sourceLayer][feature] || {};
-        extend(this.stateChanges[sourceLayer][feature], state);
+        extend(this.stateChanges[sourceLayer][feature], newState);
+
+        if (this.deletedStates[sourceLayer] === null) {
+            this.deletedStates[sourceLayer] = {};
+            for (const ft in this.state[sourceLayer]) {
+                if (ft !== feature) this.deletedStates[sourceLayer][ft] = null;
+            }
+        } else {
+            const featureDeletionQueued = this.deletedStates[sourceLayer] && this.deletedStates[sourceLayer][feature] === null;
+            if (featureDeletionQueued) {
+                this.deletedStates[sourceLayer][feature] = {};
```

Comment:
IMO using `true`/`false` to indicate "deleted"/"not deleted" would be clearer than using `null`/`{}`.

Reply:
after some thinking on this, 

1) realized that there isn't really a case for `false` here -- states not deleted simply wouldn't appear in the object
2) `true` doesn't have apparent benefits over `null`, and seems more liable to confusion with an actual state set to `true`
3) will add comment explaining the use of null, to further elucidate the usage